src/stream/xml_reader.rs

```diff
@@ -1,221 +1,262 @@
+                            let mut encoded = self.read_content(buffer)?;
                             // Strip whitespace and line endings from input string
-                            s.retain(|c| !c.is_ascii_whitespace());
-                            let data = base64::decode(&s)
+                            encoded.retain(|c| !c.is_ascii_whitespace());
+                            let data = base64::decode(&encoded)
                                 .map_err(|_| self.with_pos(ErrorKind::InvalidDataString))?;
                             return Ok(Some(Event::Data(data.into())));
                         }
-                        "date" => {
-                            let s = self.read_content()?;
-                            let date =
-                                Date::from_xml_format(&s).map_err(|e| self.with_pos(e.into()))?;
+                        b"date" => {
+                            let s = self.read_content(buffer)?;
+                            let date = Date::from_xml_format(&s)
+                                .map_err(|_| self.with_pos(ErrorKind::InvalidDateString))?;
                             return Ok(Some(Event::Date(date)));
                         }
-                        "integer" => {
-                            let s = self.read_content()?;
+                        b"integer" => {
+                            let s = self.read_content(buffer)?;
                             match Integer::from_str(&s) {
                                 Ok(i) => return Ok(Some(Event::Integer(i))),
                                 Err(_) => {
                                     return Err(self.with_pos(ErrorKind::InvalidIntegerString))
                                 }
                             }
                         }
-                        "real" => {
-                            let s = self.read_content()?;
-                            match f64::from_str(&s) {
+                        b"real" => {
+                            let s = self.read_content(buffer)?;
+                            match s.parse() {
                                 Ok(f) => return Ok(Some(Event::Real(f))),
                                 Err(_) => return Err(self.with_pos(ErrorKind::InvalidRealString)),
                             }
                         }
-                        "string" => return Ok(Some(Event::String(self.read_content()?.into()))),
+                        b"string" => {
+                            return Ok(Some(Event::String(self.read_content(buffer)?.into())))
+                        }
+                        _ => return Err(self.with_pos(ErrorKind::UnknownXmlElement)),
+                    }
+                }
+                XmlEvent::Empty(name) => {
+                    match name.local_name() {
+                        b"true" => return Ok(Some(Event::Boolean(true))),
+                        b"false" => return Ok(Some(Event::Boolean(false))),
+
+                        b"array" | b"dict" => {
```

Comment:
It looks like using [Reader::expand_empty_elements](https://docs.rs/quick-xml/latest/quick_xml/reader/struct.Reader.html#method.expand_empty_elements) can avoid us having to duplicate logic between `<array>...</array>` and `<array />` events.

Reply:
:+1: